### PR TITLE
Larkin/games

### PR DESCRIPTION
**Pull request by larkz**
- Added 2048 implementation based off Rosetta Code implementation - further edits required.
- Removed IDistribution for GridWorld MDP

### CODE REVIEW

File: lib/src/main/kotlin/MDP.kt [hunk 1]
```diff
@@ -1,9 +1,11 @@
 abstract class MDP<TState, TAction> {
-    abstract fun transition(state: TState, action: TAction) : IDistribution<TState>
+
+    abstract fun transition(state: TState, action: TAction) : TState
```
Comment (JunTaoLuo) [inline lib/src/main/kotlin/MDP.kt:3]: Why did these change? We should support probabilistic outcomes. If the game isn't probabilistic, it should just return a distribution with one single element.

File: app/src/main/kotlin/Twenty48/Game2048.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    fun shiftCellsLeft(grid: Array<Array<Int>>): Array<Array<Int>> =
+        grid.map(::mergeAndOrganizeCells).toTypedArray()
+
+    fun shiftCellsRight(grid: Array<Array<Int>>): Array<Array<Int>> =
+        grid.map { row -> mergeAndOrganizeCells(row.reversed().toTypedArray()).reversed().toTypedArray() }.toTypedArray()
+
+    fun shiftCellsUp(grid: Array<Array<Int>>): Array<Array<Int>> {
+        val rows: Array<Array<Int>> = arrayOf(
+            arrayOf(grid[0][0], grid[1][0], grid[2][0], grid[3][0]),
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048.kt:92]: I'm not a fan of hardcoding these. What if I wanted to make a 5x5 grid game. Algos shouldn't be fixed to a specific parameter that's not passed in or is a member variable.

File: lib/src/main/kotlin/MDP.kt (hunk 1 above)
Comment (larkz) [reply lib/src/main/kotlin/MDP.kt:3]: As discussed last meeting, we should aim to reduce the burden of implementing additional classes on the game side to implement, as MCTS does not explicitly require the probability class at the moment.

File: app/src/main/kotlin/Twenty48/Game2048.kt (hunk 2 above)
Comment (larkz) [reply app/src/main/kotlin/Twenty48/Game2048.kt:92]: This implementation is from [Rosetta Code](https://rosettacode.org/wiki/2048) actually. They hardcoded in the Grid, and it could require an entire rewrite to refactor. The Scala and Java implementations seem modular, we could use those too I guess.

File: app/src/main/kotlin/Twenty48/Game2048MDP.kt [hunk 3]
```diff
@@ -4,41 +4,51 @@ import MDP
...
-        // Fix issue with constant mutation, the state needs to be tied to the grid!!
-        // val newGrid = this.gameObject.run2048Programatic(this.gameObject.grid, action.toString())
-
-        // return Game2048State(newGrid, this.gameObject.gameScore.toDouble())
-        // gridworld score also
-        return Game2048State(gameObject.grid)
+
+        var tempGrid = gameObject.manipulateGrid(state.gameGrid, action.toString()) // this code mutates initialGameState
+        tempGrid = gameObject.spawnNumber(tempGrid)
+        val newGamePosition = Game2048Position(tempGrid)
```
Comment (larkz) [inline app/src/main/kotlin/Twenty48/Game2048MDP.kt:30]: This code mutates the root stateNode in the solver for some reason. It no longer goes back to the original grid of all zeroes, but always continues from the last grid configuration. @JunTaoLuo

File: app/src/main/kotlin/Mcts/MCTSSolver.kt [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            return stateNode
+        }
+
+        // Expand an unexplored action
+        val unexploredActions = stateNode.validActions.minus(stateNode.children.map { c -> c.action })
+        val actionTaken = unexploredActions.random()
+        val actionNode = createActionNode(stateNode, actionTaken)
+
+        // Transition to new state for given action
+        val newState = mdp.transition(stateNode.state, actionTaken)//.randomElement(random)
```
Comment (larkz) [inline app/src/main/kotlin/Mcts/MCTSSolver.kt:153]: To be specific, the rootNode and initialStateNode of the 2048 game is mutating after this transition. This is totally perplexing why... still investigating.

**Comment by larkz** [comment]
https://github.com/JunTaoLuo/KotlinMCTS/pull/5/commits/3569ff26dfce3a653b6f91d3f0f9f03e465c2e74#diff-374ea6c885c598960c2987423cf259e37ed064128ce6de57030bf544aa35752aR30
@JunTaoLuo here is where the mutation to the root state node happened. By creating a completely separate copy element-by-element of the grid I was able to resolve it.

File: app/src/main/kotlin/GridWorld/GridworldMDP.kt [hunk 5]
```diff
@@ -35,38 +41,39 @@ class GridworldMDP(
...
+            return GridworldState(state.x, state.y, true)
         }
 
         // if target is out of bounds, return current state
         val targetNeighbour = state.ResolveNeighbour(action, xSize, ySize) ?:
-            return UniformDistribution(listOf(state))
+            return state
 
-        val allNeighbours = mutableListOf<ProbabilisticElement<GridworldState>>()
+        val allNeighbours = mutableListOf<GridworldState>()
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/GridWorld/GridworldMDP.kt:56]: nit: call this nontargetNeighbours since it contains the possible results other than the target action.

File: app/src/main/kotlin/GridWorld/GridworldMDP.kt [hunk 6]
```diff
@@ -35,38 +41,39 @@ class GridworldMDP(
...
         for (a in GridworldAction.values().toList().minus(action)) {
             val possibleNeighbour = state.ResolveNeighbour(a, xSize, ySize)
             if (possibleNeighbour != null) {
-                allNeighbours.add(ProbabilisticElement(possibleNeighbour, 0.0))
+                allNeighbours.add(possibleNeighbour)
             }
         }
 
         // compute probability of going into non-target neighbour state
         val otherProbability = (1 - transitionProbability) / allNeighbours.size
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/GridWorld/GridworldMDP.kt:66]: This is not used anymore, remove it along with the comment above.

File: app/src/main/kotlin/GridWorld/GridworldMDP.kt [hunk 7]
```diff
@@ -35,38 +41,39 @@ class GridworldMDP(
...
 
-        for (n in allNeighbours) {
-            n.probability = otherProbability
-        }
-
         // add target neighbour
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/GridWorld/GridworldMDP.kt:68]: remove the comments here.

File: app/src/main/kotlin/GridWorldMain.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import GridWorld.GridWorldGridSolve
+import GridWorld.GridworldReward
+import GridWorld.GridworldState
+import kotlin.random.Random
+
+fun main() {
+    /*
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/GridWorldMain.kt:7]: Remove these comments.

File: app/src/main/kotlin/Twenty48/Game2048.kt [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    var gameScore: Int = 2
+
+    var grid = this.spawnNumber(arrayOf(
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0)
+    ))
+
+    // fun main(args: Array<String>) {
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048.kt:19]: remove comment.

File: app/src/main/kotlin/Twenty48/Game2048Action.kt [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package Twenty48
+
+enum class Game2048Action {
+    up, down, left, right;
+
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048Action.kt:5]: remove empty line.

File: app/src/main/kotlin/Twenty48/Game2048Controller.kt [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    fun isGridSolved(grid: Array<Array<Int>>): Boolean = grid.any { row -> row.contains(2048) }
+    fun isGridFull(grid: Array<Array<Int>>): Boolean = grid.all { row -> !row.contains(0) }
+
+    fun spawnNumber(grid: Array<Array<Int>>): Array<Array<Int>> {
+
+        val newGrid = grid.copyOf()
+        val coordinates = locateSpawnCoordinates(newGrid)
+        val number = generateNumber()
+        // val number = 2
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048Controller.kt:13]: What is this about?

File: app/src/main/kotlin/Twenty48/Game2048Controller.kt [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                if (cell == 0) emptyCells.add(Pair(x, y))
+            }
+        }
+        return emptyCells[(Math.random() * (emptyCells.size - 1)).toInt()]
+    }
+
+    fun generateNumber(): Int = if (Math.random() > 0.10) 2 else 4
+
+    fun updateGrid(grid: Array<Array<Int>>, at: Pair<Int, Int>, value: Int): Array<Array<Int>> {
+        // val updatedGrid = grid.copyOf()
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048Controller.kt:30]: remove comment.

File: app/src/main/kotlin/Twenty48/Game2048MDP.kt [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+    val initialGameState = Game2048State(Game2048Position(arrayOf(
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0),
+        arrayOf(0, 0, 0, 0)
+    )))
+
+    override fun initialState(): Game2048State {
+
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048MDP.kt:15]: remove empty line. and comment on the next line

File: app/src/main/kotlin/Twenty48/Game2048Position.kt [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package Twenty48
+
+open class Game2048Position(grid: Array<Array<Int>>) {
+    val grid = grid
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048Position.kt:4]: Why is this necessary?

File: app/src/main/kotlin/Twenty48/Game2048State.kt [hunk 15]
```diff
@@ -0,0 +1,25 @@
+package Twenty48
+
+import Twenty48.Game2048
+
+// open class Game2048State(gameObject: Game2048) {
```
Comment (JunTaoLuo) [inline app/src/main/kotlin/Twenty48/Game2048State.kt:5]: Clean up all the commented code in this class, if they are not used, remove them.

File: lib/src/main/kotlin/MCTSSolver.kt [hunk 16]
```diff
@@ -0,0 +1,283 @@
+import kotlin.math.ln
```
Comment (JunTaoLuo) [inline lib/src/main/kotlin/MCTSSolver.kt:1]: Is this class needed anymore? Compare this with the StatefulSolver and see if there are anything different. Afterwards, we should probably remove this.

File: lib/src/main/kotlin/StatefulSolver.kt [hunk 17]
```diff
@@ -24,7 +24,7 @@ class StatefulSolver<TState, TAction>(
     }
 
     fun initialize() {
-        val initialState = mdp.initialState().randomElement(random)
+        val initialState = mdp.initialState()//.randomElement(random)
```
Comment (JunTaoLuo) [inline lib/src/main/kotlin/StatefulSolver.kt:27]: Remove these comments, applies to multiple lines in this file.

File: lib/src/main/kotlin/StatelessSolver.kt [hunk 18]
```diff
@@ -101,7 +101,7 @@ class StatelessSolver<TState, TAction>(
         while(true) {
             val validActions = mdp.actions(currentState)
             val randomAction = validActions.toList().random()
-            val newState = mdp.transition(currentState, randomAction).randomElement(random)
+            val newState = mdp.transition(currentState, randomAction)//.randomElement(random)
```
Comment (JunTaoLuo) [inline lib/src/main/kotlin/StatelessSolver.kt:104]: Clean up these too. just remove the comment.

**Review by JunTaoLuo — COMMENTED** [review]
Looks pretty good

**Comment by larkz** [comment]
@JunTaoLuo all comments should be addressed.

**Comment by JunTaoLuo** [comment]
hmm? I still see quite a few that needs to be addressed. I think Github is hiding some of them because there were too many. You have to click "Show hidden comments"
